projects/feature-activation/0001-feature-activation.md

```diff
@@ -0,0 +1,558 @@
+                return FeatureState.ACTIVE
+
+            count = get_bit_count(block, feature)
+
+            if height < criteria.timeout_height and count >= criteria.threshold and height >= criteria.minimum_activation_height:
+                return FeatureState.ACTIVE
+
+            return FeatureState.STARTED
+
+        case FeatureState.ACTIVE:
+            return FeatureState.ACTIVE
+
+        case FeatureState.FAILED:
+            return FeatureState.FAILED
+```
+
+The `get_bit_count()` function is left unimplemented, together with other necessary internal functions that were used above and are defined below:
+
+```python
+def get_bit_count(block: Block, feature: Feature) -> int:
+    # return the count of enabled bits in the previous evaluation interval for this feature.
+    # walk up the blockchain using block.get_block_parent() until reaching the interval boundary,
+    # and then check if the feature bit is enabled for each block.
+    # The count must not include the bit of the block itself, only of its ancestors
+    # in the previous evaluation interval.
+    raise NotImplementedError()
+
+def _get_block(vertex: BaseTransaction) -> Block:
+    # return this vertex if it's a block, or its closest block parent if it is a transaction.
+    raise NotImplementedError()
+
+def _get_ancestor_at_height(block: Block, height: int) -> Block:
+    # return the block with the given height by walking up the
+    # blockchain using block.get_block_parent().
+    # There may be more optimized ways to do this using the height index.
+    # This is left as an unresolved implementation detail.
+    raise NotImplementedError()
+```
+
+### Caching mechanism
+
+Following the implementation above would make the `get_state()` function call itself recursively until reaching genesis. To prevent this and make the code run efficiently, a state caching mechanism is required.
+
+The state of a block/feature combination is completely determined by the state of the first block in an evaluation period, that is, the previous block with a height that is a multiple of `EVALUATION_INTERVAL`. By caching every one of those interval boundary blocks for each feature, the problem is resolved.
+
+Given the defined `EVALUATION_INTERVAL` value of `40320`, and the current blockchain height of approximately 3,400,000 blocks, the cache will store less than 100 boundary block states for each feature, and that number will be small for years.
+
+Therefore, it should be enough to store this cache in memory. It would need to be reconstructed every time the full node is restarted. If during implementation tests this reconstruction is deemed to be too slow, RocksDB may be used to permanently store it.
+
+### Dealing with reorgs
+
+When a reorg happens, re-computation of states may be necessary. The cache must be voided for every cached block that participated in the reorg.
```

Comment:
I missed more details about how reorgs should be handled. I'm assuming that each blockchain would have its own set of features. In order words, even if a given feature was activated, a reorg might disable that feature (and vice-versa).

Reply:
Indeed. I added some more explanation on https://github.com/HathorNetwork/rfcs/pull/51/commits/b43a33cd34037aa3cc625c5b77ae99b407f9fad3. Let me know if you're still missing more details.

However, I do believe there's some low-level ambiguity left open here, that may only be clearer during implementation. I've added one more dev day to the reorg task to factor that in.

Reply:
Good explanation. Thanks! I think that a simpler implementation would use metadata to store information for features, so you don't have the trouble to handle with this maybe-too-complex cache policy.

Reply:
Updated as explained in https://github.com/HathorNetwork/rfcs/pull/51#discussion_r1183923947

Reply:
Ok. Thanks!